### PR TITLE
[PLT-8024] Support LOGIN authentication method for SMTP

### PR DESCRIPTION
**Pull request by cpanato**
#### Summary
Add support for LOGIN authentication for SMTP

@jasonblais  I need some help to test the email/smtp config in other providers. I tested using microsoft outlook and with no authentication using our inbucket.

we need to test using gmail and maybe other providers. can you help me on this to validate?
thanks

#### Ticket Link
github: https://github.com/mattermost/mattermost-server/issues/7881
JIRA: https://mattermost.atlassian.net/browse/PLT-8024

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by cpanato** [comment]
the idea to make tests is updating the `inbucket` to add support for authentication. I will try to do that in the inbucket repo.

**Comment by jasonblais** [comment]
@hmhealey Would you know of a good way to test this?

**Comment by cpanato** [comment]
@jasonblais @hmhealey  I tested in the outlook/office365 and now it works, I guess we should spin one server and test gmail/mailgun/etcd....

**Comment by jasonblais** [comment]
I've generated a test server @cpanato :)

**Comment by jasonblais** [comment]
@cpanato Were you going to test it via the spinmint?

**Comment by cpanato** [comment]
yes, give to me a few more days  please. busy week :/

**Comment by jasonblais** [comment]
@cpanato No rush! I just realized I wasn't sure what the next step was :)

**Comment by jasonblais** [comment]
Certainly! No rush at all :)

File: utils/mail.go [hunk 1]
```diff
@@ -20,6 +21,32 @@ import (
...
 
+type loginAuth struct {
+	username, password string
+}
+
+func LoginAuth(username, password string) smtp.Auth {
+	return &loginAuth{username, password}
+}
+
+func (a *loginAuth) Start(server *smtp.ServerInfo) (string, []byte, error) {
```
Comment (ccbrown) [inline utils/mail.go]: I assume you'll inspect `server.Auth` so we can continue to use PLAIN where LOGIN isn't supported?

We should also continue to error out on `!server.TLS` for remote servers.

Comment (cpanato) [reply utils/mail.go]: ok sure!

**Comment by cpanato** [comment]
@jasonblais removed the server found some issues. will work using my local dev.

**Comment by cpanato** [comment]
tested with in my local
 - smtp.mailgun.org - Mailgun using ports: 25 (none), 465 (TLS) and 587(StartTLS)
 - smtp-mail.outlook.com - Outlook ports 25(StartTLS) and 587(StartTLS)
 - smtp.gmail.com - Gmail ports: 587(StartTLS), 465(TLS)

**Comment by cpanato** [comment]
mailgun looks like PLAIN auth and outlook uses LOGIN, both we can use the code above and works 

I tried to use the existing implementation and it failed for outlook but works for mailgun. using this code in the PR works for both
@ccbrown

**Comment by cpanato** [comment]
@ccbrown I will ping you in the channel to sync with you. if is that ok

**Comment by ccbrown** [comment]
Of course it’s okay. But I’m going to sleep so I won’t be around for a while. :P 

My main concern is that removing support for PLAIN auth would be a breaking change. I don’t know if any of our customers use SMTP servers that don’t support LOGIN, but it’s certainly possible.

There’s also the issue of what to do for servers that support both. Which is better? LOGIN or PLAIN? Turns out LOGIN is obsoleted, and the authentication method we should use is actually well defined:

> Clients SHOULD implement the PLAIN SASL mechanism and use it whenever
    offered by a server.  The LOGIN SASL mechanism SHOULD NOT be used by
    a client when other plaintext mechanisms are offered by a server.

– https://tools.ietf.org/html/draft-murchison-sasl-login-00

**Comment by cpanato** [comment]
I push some changes that I would like to test in the test server, this is not the end code.

**Comment by cpanato** [comment]
tested in the test server:

* smtp.mailgun.org - Mailgun using ports: 25 (none), 465 (TLS) and 587(StartTLS)
* smtp-mail.outlook.com - Outlook ports 25(StartTLS) and 587(StartTLS)
* smtp.gmail.com - Gmail ports: 587(StartTLS), 465(TLS)

i will try to refactor a bit the code. and then submit for review

**Comment by cpanato** [comment]
all the merits goes to @ccbrown

**Comment by cpanato** [comment]
@jasonblais can you spin a new test server? and i will do some tests and will be ready to go

**Comment by jasonblais** [comment]
@cpanato Done :)

**Comment by cpanato** [comment]
tested with `gmail` and `mailgun` and works fine.

using `outlook.com` it failed in the test server but in my local development it works fine.

I'm getting this error: `Connection unsuccessful: Failed to close connection to SMTP server` 
and this part is when it try to send the email to the server and try to close the connection: https://github.com/mattermost/mattermost-server/blob/master/utils/mail.go#L205

the logs says:
```
[2018/02/08 13:59:58 UTC] [EROR] /api/v4/email/test:SendMail code=500 rid=dyr8n7q5bjgctda7b69kzs4eqe uid=tn4ndpia6b8iiyhpi7y5ijx6rc ip=89.16.153.196 Failed to close connection to SMTP server [details: 554 5.2.0 STOREDRV.Submission.Exception:SendAsDeniedException.MapiExceptionSendAsDenied; Failed to process message due to a permanent exception with message Cannot submit message. 16.55847:A8090000, 17.43559:00000000A0000000000000000000000000000000, 20.52176:140FE5810200101064010000, 20.50032:140FE5817617101069010000, 0.35180:0A002880, 255.23226:6E010000, 255.27962:7A000000, 255.27962:0A000000, 255.27962:0E000000, 255.31418:78010000, 16.55847:C2000000, 17.43559:0000000068010000000000000000000000000000, 20.52176:140FE5810200101087010000, 20.50032:140FE581761710108C010000, 0.35180:40000730, 255.23226:91010000, 255.27962:0A000000, 255.27962:32000000, 255.17082:DC040000, 0.27745:9F010000, 4.21921:DC040000, 255.27962:FA000000, 255.1494:40000830, 0.37692:B1000000, 0.37948:B1000000, 5.33852:00000000534D545000000000, 4.56248:DC040000, 7.40748:010000000000010B30323A44, 7.57132:000000000000000038393A37, 1.63016:32000000, 4.39640:DC040000, 8.45434:FE7F03008410DE4F000000000000000000000000, 5.10786:0000000031352E32302E303438352E3030323A4442365052303530314D42313938393A37316238666266342D346564642D343464642D613665372D3362666662353939656363660000000000, 255.1750:41010000, 255.31418:0A002080, 0.22753:46010000, 255.21817:DC040000, 4.60547:DC040000, 0.21966:0A002280, 4.30158:DC040000 [Hostname=DB6PR0501MB1989.eurprd05.prod.outlook.com]]
[2018/02/08 13:59:58 UTC] [EROR] /api/v4/logs:client code=0 rid=zsqma1ig378tzq1t9mox33b68w uid=tn4ndpia6b8iiyhpi7y5ijx6rc ip=89.16.153.196 {"message":"Failed to close connection to SMTP server","server_error_id":"utils.mail.send_mail.close.app_error","status_code":500,"url":"http://i-07df579ac15af940e.spinmint.com/api/v4/email/test"} [details: ]
```
I found this answer: https://social.technet.microsoft.com/Forums/office/en-US/69426f64-717d-47e0-aa1b-700903ee83cc/520-storedrv-error?forum=exchangesvrsecuremessaginglegacy

and if I change the address `Notification From Address:` and set the same email as we use in the smtp user it works

**Comment by cpanato** [comment]
after the configuration using `smtp-mail.outlook.com` the test user was able to receive email from mention notification

**Comment by jasonblais** [comment]
@cpanato Huge thanks for working on this! Notes:

1) @lindalumitchell will help test Amazon SES, with our usual email test credentials.
2) You tested `smtp.gmail.com` and `smtp.mailgun.org`, correct? And you got `smtp-mail.outlook.com` to work as expected? 
3) Should we also test on [`localhost`](https://docs.mattermost.com/install/smtp-email-setup.html?highlight=smtp#postfix)?
4) One purpose of adding LOGIN auth support is to support OWA (Outlook Web App). Wondering if we've tested this yet?
5) Do we need [documentation changes for SMTP set up](https://docs.mattermost.com/install/smtp-email-setup.html)? Looks like none?
6) How are we addressing Chris' comment above?

>My main concern is that removing support for PLAIN auth would be a breaking change. I don’t know if any of our customers use SMTP servers that don’t support LOGIN, but it’s certainly possible.
> There’s also the issue of what to do for servers that support both. Which is better? LOGIN or PLAIN? Turns out LOGIN is obsoleted, and the authentication method we should use is actually well defined:

**Comment by ccbrown** [comment]
@jasonblais for number 6, @cpanato updated the PR to use PLAIN when available, and LOGIN otherwise. So I'm happy now; no more compatibility or spec/compliance concerns from me.

**Comment by lindalumitchell** [comment]
@cpanato @jasonblais My testing looked good, using our Amazon SES test credentials. Tested transactional emails (new email, password) and various types of notifications (at-mentions/DMs, batched/unbatched, generic/including message snippet). No issues found.

**Comment by cpanato** [comment]
@jasonblais 
2. Works as expected with all transactional emails.
3. there is a test case for that? and using localhost this use any authentication or there is no auth? if there is no auth we can skip.
4. yes I tested using the outlook and works

5. this is transparent to the user. the actual documentation covers all.

6. all the merits go to @ccbrown he helped me a lot on this PR

**Comment by jasonblais** [comment]
3. That's correct, localhost is without auth (SMTP settings are here https://docs.mattermost.com/install/smtp-email-setup.html#postfix).

I think this is good for dev review then! Big thanks for your help :)

**Comment by cpanato** [comment]
@ccbrown I will think about how to add some tests.

**Comment by cpanato** [comment]
@ccbrown  @hmhealey  yes I agree, I need some help in how to make it.

**Comment by cpanato** [comment]
@ccbrown @hmhealey I added an initial unit test to check the AUTH thing. For the server itself, I might need more time to check how to implement. Maybe we can talk in Lisbon about this

**Comment by cpanato** [comment]
@ccbrown @hmhealey my idea is to implement the authentication in the inbucket side (for my personal development :-) )
I'm gathering some resources and knowledge to do that. For Now i don't have any estimates

**Comment by jwilander** [comment]
We can merge this right after the release branch gets cut

**Comment by jwilander** [comment]
@cpanato one conflict to fix and feel free to merge :)

**Comment by cpanato** [comment]
rebased and merging

**Pull request merged**